discovery-provider/src/solana/solana_program_indexer.py

```diff
@@ -0,0 +1,263 @@
+import asyncio
+import logging
+from abc import ABC, abstractmethod
+from typing import Any, Dict
+
+from src.solana.constants import (
+    TX_SIGNATURES_BATCH_SIZE,
+    TX_SIGNATURES_MAX_BATCHES,
+    TX_SIGNATURES_RESIZE_LENGTH,
+)
+from src.solana.solana_transaction_types import TransactionInfoResult
+
+logger = logging.getLogger(__name__)
+
+BASE_ERROR = "Must be implemented in subclass"
+
+TX_SIGNATURES_PROCESSING_SIZE = 100
+
+
+class IndexerBase(ABC):
+    """
+    Base indexer class - handles lock acquisition to guarantee single entrypoint
+    """
+
+    _label: str
+    _lock_name: str
+    _redis: Any
+    _db: Any
+
+    @abstractmethod
+    def __init__(self, label: str, redis: Any, db: Any):
+        self._label = label
+        self._redis = redis
+        self._db = db
+        self._lock_name = f"{self._label}"
```

Comment:
Maybe prefix all the solana program indexer lock names ie `f"indexer:solana:{self._label}"`